tests/integration/utils.py

```diff
@@ -162,86 +156,91 @@ def __get_reference_type(object_path, root_object_paths=None):
         # Check resource attributes
         actual_attributes = resource['attributes']
         expected_attributes = __get_schema_attributes()
-        __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed)
+        __check_attributes_schema(actual_attributes, expected_attributes)
 
     # Helper function to check error object schema
-    def __check_error_schema(error, null_value_allowed):
+    def __check_error_schema(error):
         # Check error attributes
         actual_attributes = error
         expected_attributes = schema
-        __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed)
+        __check_attributes_schema(actual_attributes, expected_attributes)
 
     # Helper function to check through all attributes
-    def __check_attributes_schema(actual_attributes, expected_attributes,
-                                  null_value_allowed=False):
+    def __check_attributes_schema(actual_attributes, expected_attributes):
         for field, actual_value in actual_attributes.items():
             expected_attribute = expected_attributes[field]
             expected_type = __get_attribute_type(expected_attribute)
-            if not null_value_allowed or (actual_value and expected_type):
-                self.assertIsInstance(actual_value, expected_type)
+            if (actual_value and expected_type) or (
+              'nullable' in expected_attribute and
+               expected_attribute['nullable'] is False):
+                    self.assertIsInstance(actual_value, expected_type)
 
     status_code = response.status_code
     content = get_json_content(self, response)
 
-    # TODO: Add self-link testing
     # Basic tests for successful/error response
     try:
         if status_code == 200:
             resource_data = content['data']
             if isinstance(resource_data, list):
                 for resource in resource_data:
-                    __check_resource_schema(resource, null_value_allowed)
+                    __check_resource_schema(resource)
             else:
-                __check_resource_schema(resource_data, null_value_allowed)
+                __check_resource_schema(resource_data)
         elif status_code >= 400:
             errors_data = content['errors']
             self.assertIsInstance(errors_data, list)
             for error in errors_data:
-                __check_error_schema(error, null_value_allowed)
+                __check_error_schema(error)
     except KeyError as error:
         self.fail(error)
 
 
 # Check url for correct base and endpoint, parameters
 def check_url(self, actual_url, endpoint, query_params=None):
+    if query_params is None:
+        query_params = {}
+
+    base_url = self.base_url
+    if self.local_test:
+        base_url = re.sub('(:\d{4}/api)', '', self.base_url)
```

Comment:
Is the parenthesis needed for the regex pattern? Also, can you add a docstring here to describe the purpose of this line?